karapace/master_coordinator.py

```diff
@@ -49,16 +50,25 @@ def group_protocols(self):
 
     def _perform_assignment(self, leader_id, protocol, members):
         self.log.info("Creating assignment: %r, protocol: %r, members: %r", leader_id, protocol, members)
-        self.master = None
+        self.are_we_master = None
         error = NO_ERROR
         urls = {}
+        fallback_urls = {}
         for member_id, member_data in members:
```

Comment:
Suggestion, I think this is a bit more straightforward:

```python
choose_highest = self.election_strategy.lower() == "highest"

# Make sure master eligible nodes are in front of the *sorted* list
def rank(member: dict) -> int:
    if member["master_eligibility"] is True:
        return 1 if choose_highest else -1
    return 0

members_parsed_data = [
    (member_id, json.loads(member_data.decode("utf8")))
    for member_id, member_data in members
]
members_with_rank = [
    (
        rank(member_data),
        member_identity["scheme"],
        member_identity["host"],
        member_identity["port"],
        member_id,
        member_data,
    )
    for member_id, member_data in members_parsed_data
]
master = sorted(members_with_rank, reverse=choose_highest)[0]
master_identity = self.get_identity(
    host=master[2],
    port=master[3],
    scheme=master[1],
    json_encode=False,
)
```

Reply:
Note: Perhaps instead of having `rank` like we have above, we should instead filter out the nodes that have `member["master_eligibility"] == False`

Reply:
To be honest, I find my code more readable, to be concrete your `rank` function has double negation for `choose_highest`(first in rank, then reverse flag for sort).

And the beef in this PR is anyway handle the case where filtering out `member["master_eligibility"] == False` lead to zero length list, and that case must be handled in one way or another.

Reply:
> To be honest, I find my code more readable, to be concrete your rank function has double negation for choose_highest(first in rank, then reverse flag for sort).

Fair enough, I don't like that myself. Just to clarify, I think of simpler as code that uses the fewest concepts and that is a bit more declarative. And in this case I was trying to reduce the number of states in the function (the if/else statements scattered all around).